### PR TITLE
ARROW-14731: [C++][Compute] Conversion to unsigned types can be problematic for negative numbers

### PR DESCRIPTION
**Pull request by bkmgit**
revert to signed 32 bit type

### CODE REVIEW

**Comment by pitrou** [comment]
Is this fixing a particular bug? If so, I'd expect a test to be added.

**Comment by bkmgit** [comment]
@pitrou An example following https://stackoverflow.com/questions/1855459/maximum-value-of-int is below:
```c++
#include<iostream>
#include<limits>
int main()
{
        int imin = std::numeric_limits<int>::min();
        int imax = std::numeric_limits<int>::max();
        int ia = 105;
        int ib = -2;
        int c;
        // Typically range of unsigned int is [0,4294967295]
        unsigned int d;
        // Typically range of int is [-2147483648,2147483647]
        int ba = imax-1;
        int bb = -2;
        int bc = 2;
        int16_t i16a = std::numeric_limits<int16_t>::max() -1;
        int i16a_32 = std::numeric_limits<int16_t>::max() -1;
        int16_t i16b = -2;
        int16_t i16c;
        uint16_t i16d;
        // Do some multiplication
        c = ia * ib;
        std::cout << c << std::endl;
        // answer should be -210

        // cast to unsigned and do multiplication
        c = (int) (((unsigned int) ia ) * ((unsigned int) ib ));
        std::cout << c << std::endl;
        // answer should be -210 after casting again

        // cast to unsigned and do multiplication
        c = (int) (((unsigned int) ba) * ((unsigned int) bb ));
        std::cout << c << std::endl;
        // Cannot represent answer -4294967292 get overflow and output 4
        
        // cast to unsigned and do multiplication
        d = (unsigned int) (((unsigned int) ba) * ((unsigned int) bb ));
        std::cout << d << std::endl;
        // output should be -4294967292 but cannot represent it so get output 4

        // cast to unsigned and do multiplication
        c = (int) (((unsigned int) ba) * ((unsigned int) bc ));
        std::cout << c << std::endl;
        // output should be 4294967292 which cannot be represented and get -4

        // cast to unsigned and do multiplication
        d = (unsigned int) (((unsigned int) ba) * ((unsigned int) bc ));
        std::cout << d << std::endl;
        // output should be 4294967292

        // cast to unsigned and do multiplication
        c = (int) (((unsigned int) ba) * ((unsigned int) bc ));
        std::cout << c << std::endl;
        // output should be 4294967292 which cannot be represented and get -4

        // cast to unsigned and do multiplication
        c = (int) ((( int) i16a) * (( int) i16b ));
        std::cout << c << std::endl;
        // output should be -65532

        // cast to unsigned int and do multiplication
        i16c = (int16_t) (((unsigned int) i16a ) * ((unsigned int) i16b ));
        std::cout << i16c << std::endl;
        // output should be -65532 which cannot be represented and get 4

        // cast to unsigned int and do multiplication
        i16d = (uint16_t) (((unsigned int) i16a ) * ((unsigned int) i16b ));
        std::cout << i16d << std::endl;
        // output should be -65532 which cannot be represented and get 4 

        return 0;
}
```
The ranges for the types do depend on the machine being used and also the programming language, but for C++, a minimum range is guaranteed by the standard. Can turn these into tests and put them in

**Comment by bkmgit** [comment]
Some examples for how to work with overflows are available at https://wiki.sei.cmu.edu/confluence/display/c/INT32-C.+Ensure+that+operations+on+signed+integers+do+not+result+in+overflow
The basic idea is to use the limits to check for overflows and report an error if there is one. Int types do not have NAN or other error representation, so would probably be good to output a warning/error but let the computation continue.

File: cpp/src/arrow/compute/kernels/scalar_arithmetic.cc [hunk 1]
```diff
@@ -247,12 +247,14 @@ struct Multiply {
...
-  // well defined.
+  // behaviour). One could first cast to 32 bit unsigned integers where overflow is
+  // well defined, but this is only reasonable if both numbers are positive or 
+  // both numbers are negative. It may be better for the user to check for an overflow
+  // or have a slow safe computation in which the program checks for an overflow
   template <typename T, typename Arg0, typename Arg1>
   static constexpr enable_if_same<T, int16_t, T> Call(KernelContext*, int16_t left,
                                                       int16_t right, Status*) {
-    return static_cast<uint32_t>(left) * static_cast<uint32_t>(right);
+    return static_cast<int32_t>(left) * static_cast<int32_t>(right);
```
Comment (cyb70289) [inline cpp/src/arrow/compute/kernels/scalar_arithmetic.cc:257]: Looks to me `uint32` or`int32` doesn't matter here.
Actually I think the even the casting to 32 bit integer is not necessary: `int16 * int16` won't overflow an `int32`.
Only `uint16 * uint16` may overflow `int32` and trigger UB.

**Comment by cyb70289** [comment]
@bkmgit , are you still working on this?
Arithmetic kernels have checked version which reports overflow [1]. Looks this PR is not necessary?

[1] https://github.com/apache/arrow/blob/master/cpp/src/arrow/compute/kernels/scalar_arithmetic.cc#L146

**Comment by bkmgit** [comment]
Thanks for letting me know. Checking code. Can then close this. Conversion is still the same:
https://github.com/apache/arrow/blob/master/cpp/src/arrow/compute/kernels/scalar_arithmetic.cc#L252
Would expect a cast to int32 of the final result so that multiplication of a positive and negative number is correct, but maybe this is done elsewhere, as the test passes:
https://github.com/apache/arrow/blob/master/cpp/src/arrow/compute/kernels/scalar_arithmetic_test.cc#L818

**Comment by cyb70289** [comment]
Close this PR. Feel free to reopen if new findings.

**Pull request closed without merging**